### ISSUE TITLE
Persist anonymous volumes on service upgrades ( catalog stack upgrades )

### ISSUE THREAD

**Issue by EugenMayer** [rancher/rancher#6109]
**Rancher Version:**
1.2.0-rc3
**Docker Version:**
1.11.2
**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
- To be able to easy reproduce this, add this catalog https://github.com/EugenMayer/docker-rancher-extra-catalogs
- deploy the catalog 'volume peristence test' in the version 0.0.1
- upgrade the stack to 0.0.2 afterwards
- open a shell on the app container, check /data - only 0.0.2 exists

**Results:** 
The test image is configure to touch a file with its version on container start (see the [Dockerfile](https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/tests/volume-persistence-test/Dockerfile#L6) and the [build configuration](https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/tests/volume-persistence-test/docker-compose.yml) ), so after starting catalog:0.0.1, /data/0.0.1 is created. Since /data is exposed as a anonymous volume, this should be persisted over catalog updates, right?

Still, after upgrading the stack (catalog) from 0.0.1 to 0.0.2 you only find /data/0.0.2 - this means that the anon image has been deleted and a new one has been created

**Expected:**
/data/0.0.1 and /data/0.0.2 should exist after the stack has been upgraded to 0.0.2 - so basically anonymous volume persistence on service upgrades of containers

**Comment by EugenMayer** [comment]
This is a kinda-dupe of https://github.com/rancher/rancher/issues/5010 with easier reproduction, also for the future. This also relates / blocks a solution to https://github.com/rancher/rancher/issues/2159#issuecomment-251115190 using data-containers to share a state in a service across all scaled container.

Using named volumes is blocked by "only a single stack of this kind of this catalog per host"

**Comment by EugenMayer** [comment]
To pressure this issue a bit more, if this issue will not get fixed, every user using a 

- wordpress
- teamcity
- wekan 
- owncload
- openldap
....

catalog will lose their data entirely, when a catalog upgrade will be released.

This holds of a lot more community catalogs, i just cherry picked on some.

**Comment by alena1108** [comment]
Volume persistance support will be added in Rancher 1.2 via new Volumetemplate/volume per stack/deployment unit feature. @deniseschannon could you update the bug with the link to the doc once it is written?

**Comment by EugenMayer** [comment]
Great news on that, i am really happy to see this going this route. Thanks!

**Comment by deniseschannon** [comment]
This use case is not actually addressed.

In 1.2, we recommend using named volumes which will persist volumes.

**Comment by psyhonut** [comment]
@deniseschannon can you explain how to create data volumes in Rancher 1.2?

**Comment by deniseschannon** [comment]
@psyhonut 

http://docs.rancher.com/rancher/v1.2/en/rancher-services/storage-service/

**Comment by alena1108** [comment]
Looks like a duplicate of https://github.com/rancher/rancher/issues/5010, going to keep track of the issue there.

**Comment by EugenMayer** [comment]
@alena1108 its a dupe, but with much more substance to reproduce and test. anyway, fair enough, one ticket should do it :)
